File: swarmmanagercontroller/src/main/java/co/uk/swarmbit/docker/api/services/ServicesApiImpl.java
package co.uk.swarmbit.docker.api.services;

import co.uk.swarmbit.docker.api.common.AbstractApiImpl;
import co.uk.swarmbit.docker.api.common.json.ServiceJson;
import co.uk.swarmbit.docker.api.services.parameters.*;
import co.uk.swarmbit.docker.api.common.json.ServiceGeneralResponseJson;
import co.uk.swarmbit.rest.RestResponseType;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class ServicesApiImpl extends AbstractApiImpl implements ServicesApi {

    private static final String SERVICES_PATH = "/services";

    private static final String CREATE_PATH = "/create";

    private static final String UPDATE_PATH = "/update";

    private static final String LOGS_PATH = "/logs";

    @Override
    public List<ServiceJson> listServices(String swarmId, ServicesListParameters parameters) {
        return listObjects(SERVICES_PATH, swarmId, new RestResponseType<List<ServiceJson>>() {}, parameters);
    }

    @Override
    public ServiceJson inspectService(String swarmId, String id, ServiceInspectParameters parameters) {
        return inspectObject(SERVICES_PATH, swarmId, new RestResponseType<ServiceJson>() {}, id, parameters);
    }

    @Override
    public ServiceGeneralResponseJson createService(String swarmId, ServiceCreateParameters parameters) {
        return createObject(SERVICES_PATH + CREATE_PATH, swarmId, new RestResponseType<ServiceGeneralResponseJson>() {},
            parameters, parameters);
    }

    @Override
    public void updateService(String swarmId, String id, ServiceUpdateParameters parameters) {
        updateObject(SERVICES_PATH + "/" + id + UPDATE_PATH, swarmId, new RestResponseType<ServiceGeneralResponseJson>() {},
                parameters, parameters, parameters);
    }

    @Override
    public void deleteService(String swarmId, String id) {
          deleteObject(SERVICES_PATH + "/" + id, swarmId, new RestResponseType<Void>() {});
    }

    @Override
    public byte[] getServiceLogs(String swarmId, String id, ServiceLogsParameters parameters) {
        return getObjectLogs(SERVICES_PATH + "/" + id + LOGS_PATH, swarmId, parameters);
    }
}

File: swarmmanagercontroller-new/src/main/kotlin/com/swarmbit/docker/api/services/ServicesApiImpl.kt
package com.swarmbit.docker.api.services

import com.swarmbit.docker.api.common.AbstractApiImpl
import com.swarmbit.docker.api.common.client.DockerWebClient
import com.swarmbit.docker.api.common.rest.model.ResponseType
import com.swarmbit.docker.api.services.model.Service
import com.swarmbit.docker.api.services.model.ServiceGeneralResponse
import com.swarmbit.docker.api.services.parameters.ServicesCreateParameters
import com.swarmbit.docker.api.services.parameters.ServicesInspectParameters
import com.swarmbit.docker.api.services.parameters.ServicesListParameters
import com.swarmbit.docker.api.services.parameters.ServicesLogsParameters
import com.swarmbit.docker.api.services.parameters.ServicesUpdateParameters
import io.micronaut.context.annotation.Context

@Context
class ServicesApiImpl(dockerWebClient: DockerWebClient) : AbstractApiImpl(dockerWebClient), ServicesApi {
    companion object {
        private const val SERVICES_PATH = "/services"
        private const val CREATE_PATH = "/create"
        private const val UPDATE_PATH = "/update"
        private const val LOGS_PATH = "/logs"
    }

    override fun listServices(swarmId: String, parameters: ServicesListParameters): List<Service> {
        return listObjects(SERVICES_PATH, swarmId, object : ResponseType<List<Service>>() {}, parameters)
    }

    override fun inspectService(swarmId: String, id: String, parameters: ServicesInspectParameters): Service {
        return inspectObject(SERVICES_PATH, swarmId, object : ResponseType<Service>() {}, id, parameters)
            ?: throw IllegalArgumentException("No service found for swarmId ($swarmId) and serviceId (${id})")
    }

    override fun createService(swarmId: String, parameters: ServicesCreateParameters): ServiceGeneralResponse {
        return createObject(SERVICES_PATH + CREATE_PATH, swarmId, object : ResponseType<ServiceGeneralResponse>() {}, parameters, parameters)!!
    }

    override fun updateService(swarmId: String, id: String, parameters: ServicesUpdateParameters) {
        updateObject("$SERVICES_PATH/$id$UPDATE_PATH", swarmId, object : ResponseType<ServiceGeneralResponse>() {},
            parameters, parameters, parameters)
    }

    override fun deleteService(swarmId: String, id: String) {
        deleteObject("$SERVICES_PATH/$id", swarmId, object : ResponseType<Void>() {})
    }

    override fun getServiceLogs(swarmId: String, id: String, parameters: ServicesLogsParameters): ByteArray {
        return getObjectLogs("$SERVICES_PATH/$id$LOGS_PATH", swarmId, parameters) ?: ByteArray(0)
    }

}
